### ISSUE TITLE
Deprecate kB, MB, GB, TB and introduce KiB, MiB, GiB, TiB

### ISSUE THREAD

**Issue by typingArtist** [SpinalHDL/SpinalHDL#183]
We should use and encourage proper binary prefix according to the IEC standard. The old/wrong/JEDEC-only forms should be deprecated.

https://en.wikipedia.org/wiki/Binary_prefix

Edit: Also updated such that in addition to ```Byte``` also ```Bytes``` is supported for the unit.

**Comment by tomverbeure** [comment]
> We should use and encourage proper binary prefix 

I think we should not. 

Insisting on KiB and MiB is like insisting that the new century started in 2001 and not 2000. It may be technically correct, but it’s also annoying. ;-)

**Comment by typingArtist** [comment]
I’d say, insisting on KiB and MiB is like acknowledging that a new century has started at all … :grin:

And the old form is still supported, with deprecation warnings though.

**Comment by tomverbeure** [comment]
What about a reasonable middle ground: Add KiB and MiB, but don't have deprecation warnings? There's really no upside in having those deprecation warnings.

**Comment by zhutmost** [comment]
In most academic papers about computer architecture or hardware, we still use kB instead of KiB much often. The same things also happen in lots of chip datasheets. 
I guess deprecation warnings will confuse users? 

( actually, many standards made by IEEE or other associations are insisted by only a little people. For example, SystemVerilog :D

**Comment by tomverbeure** [comment]
I’ll just leave this here: https://xkcd.com/394/. :-)

**Comment by Dolu1990** [comment]
So, keeping both kB, MB, GB, TB and KiB, MiB, GiB, TiB, without depreciation ?

**Comment by tomverbeure** [comment]
Yes!

**Comment by typingArtist** [comment]
Other projects have already found out about the old form’s ambiguity and have reacted. Why can’t we do as well? I’d like to see the old form go away better soon than late, but @Dolu1990 has already made clear that he’d like to keep the old form around with deprecation warnings.

I think that this is the best we can do, so I vote against un-deprecating the old form.

**Comment by Dolu1990** [comment]
So the thing is, maybe today this sound stupid to deprecate the KB MB stuff, but what about in the long terme ? 
As SpinalHDL is there to reign on the EDA ecosystem for the next 1000 years, it is probably the right move ^.^, or not ?

**Comment by typingArtist** [comment]
That’s why we also have YiB in the new version … and we will patch it as soon as [support for the next prefix](http://googology.wikia.com/wiki/Prefix_10%5E27) is standardized! :grinning:

**Comment by typingArtist** [comment]
I’d say we are done here. Closing.

**Comment by Dolu1990** [comment]
I don't know, Another solution, is to not reconize KB as something else than 1024
Who in hardware design will ever use KB as 1000 ?
I think nobody will ever need it as 1000. 

So still thinking about removing the depreciation on the KB, and just adding a note on the doc about it XD

**Comment by typingArtist** [comment]
Well … think of Ethernet with 1 Gbit/s. Is it Gbit or Gibit? It can transport 125 MB per second … but which MB? MB or MiB? Do we need to care? Definitely yes.

What shall we do if we ever connect SpinalHDL hardware design closer with software, drivers, configuration tools … I have just seen too many cases where the difference between MB and MiB made for bugs and mistakes on the interface between hardware and software.

**Comment by Dolu1990** [comment]
Apparently, KB is 1024, whil kB is 1000. I think this will make everybody happy !
![image](https://user-images.githubusercontent.com/4832082/51073484-bed11600-1671-11e9-891e-0971a47dc44b.png)
from 
https://en.wikipedia.org/wiki/Kibibyte

**Comment by typingArtist** [comment]
BTW, another example of bad notation is PCI and DRAM with its ⅓ of 100 MHz steps … PCI is said to be 33 MHz … is it 33 or 33.3 MHz or 33.333333 MHz, i.e. 3.0 ns cycle time? Should we care? Definitely yes.

BTW, a lot of wrong values in that table because of this … https://en.wikipedia.org/wiki/Data-rate_units#Examples_of_bit_rates

And I don’t see any compelling reason to make an exception just for KB!?

**Comment by Dolu1990** [comment]
Ahh right, for baud rate, it is more ambigus, good point ^^

**Comment by typingArtist** [comment]
Once I had a case where I needed a PLL to be configured for 266 MHz for a memory interface. Guess what, I got 266.0 MHz from the other engineer, instead of the expected 266.66…67 MHz (3.75 ns). It was my fault because of the bad habit of assuming that ‘the other side will understand what I meant anyway’.

Or the case of a memory block that was exactly 1 MB (well … 1 MiB) and it was expected to be emptied by an interface that was capable of doing 1 MB/s (well … exactly 1 MB/s) in exactly 1 s. Turns out, it did take significantly longer – actually 48.576 ms longer – breaking the memory management.
